src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,219 @@
+        );
+        // remove allocation
+        let tcx = self.ecx.tcx;
+        let memory = self.ecx.memory_mut();
+        let (kind, mut alloc) = match memory.alloc_map.remove(&ptr.alloc_id) {
+            Some(entry) => entry,
+            None => {
+                if tcx.alloc_map.lock().get(ptr.alloc_id).is_none() {
+                    tcx.sess.delay_span_bug(self.ecx.tcx.span, "tried to intern dangling pointer");
+                }
+                // treat dangling pointers like other statics
+                // just to stop trying to recurse into them
+                return Ok(Some(IsStaticOrFn));
+            },
+        };
+        match kind {
+            MemoryKind::Stack | MemoryKind::Vtable => {},
+        }
+        // ensure llvm knows not to put this into immutable memory
+        alloc.mutability = match self.mode {
+            InternMode::StaticMut => Mutability::Mutable,
+            InternMode::Static { frozen: true } => Mutability::Immutable,
+            InternMode::Static { frozen: ref mut frozen @ false } => {
+                // values behind the initial reference are frozen (except if the type undoes this)
+                *frozen = true;
+                Mutability::Mutable
+            },
+            InternMode::Const { frozen: true } => Mutability::Immutable,
+            InternMode::Const { frozen: ref mut frozen @ false } => {
+                // values behind the initial reference are frozen even if the type would undo this.
+                // validation will detect any discrepancies and tell the user that this is not cool.
+                *frozen = true;
+                Mutability::Mutable
+            },
+        };
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Go over all the primitive types
+        let ty = mplace.layout.ty;
+        if let ty::Ref(..) = ty.sty {
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait objects
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            self.intern(vtable)?;
```

Comment:
Vtables should always be read-only... is this reflected somewhere, somehow?

Reply:
It's directly behind a reference, so it will always be immutable.

Reply:
I am not sure what you mean by this. What if the reference is mutable or has interior mutability?

Can this `assert!` that `self.mode` is something that makes things immutable?

Reply:
yea that should work

Reply:
(The assert still does not seem present in the latest version.)